src/toil/provisioners/clusterScaler.py

```diff
@@ -84,150 +85,180 @@ def get(self):
-            consideredNodes = [nodeShape for nodeShape in nodeShapes if nodeShape.cores >= jS.cores and nodeShape.memory >= jS.memory and nodeShape.disk >= jS.disk and (jS.preemptable or not nodeShape.preemptable)]
-            if len(consideredNodes) == 0:
-                return
-            nodeShape = consideredNodes[0]
-            x = NodeReservation(subtract(nodeShape, jS))
-            nodeReservations[nodeShape].append(x)
-            t = nodeShape.wallTime
-            while t < jS.wallTime:
-                y = NodeReservation(x.shape)
-                t += nodeShape.wallTime
-                x.nReservation = y
-                x = y
-
-        addToReservation()
-    return {nodeShape:len(nodeReservations[nodeShape]) for nodeShape in nodeShapes}
+        jobShape is the Shape of the job requirements, nodeShape is the Shape of the node this
+        is a reservation for, and targetTime is the maximum time to wait before starting this job.
+        """
+        # starting slice of time that we can fit in so far
+        startingReservation = self
+        # current end of the slices we can fit in so far
+        endingReservation = startingReservation
+        # the amount of runtime of the job currently covered by slices
+        jobTimeSoFar = 0
+        # total time from when the instance started up to startingReservation
+        startingReservationTime = 0
+
+        while True:
+            # Considering a new ending reservation.
+            if endingReservation.fits(jobShape):
+                jobTimeSoFar += endingReservation.shape.wallTime
+
+                if jobTimeSoFar >= jobShape.wallTime:
+                    # The job fits into all the slices between startingReservation and endingReservation.
+                    t = 0
+                    # Update all the slices, reserving the amount of resources that this job needs.
+                    while startingReservation != endingReservation:
+                        startingReservation.subtract(jobShape)
+                        t += startingReservation.shape.wallTime
+                        startingReservation = startingReservation.nReservation
+                    assert startingReservation == endingReservation
+                    assert jobShape.wallTime - t <= startingReservation.shape.wallTime
+                    adjustEndingReservationForJob(endingReservation, jobShape, t)
+                    # Packed the job.
+                    return True
+
+                # If the job would fit, but is longer than the total node allocation
+                # extend the node allocation
+                elif endingReservation.nReservation == None and startingReservation == self:
+                    # Extend the node reservation to accommodate jobShape
+                    endingReservation.nReservation = NodeReservation(nodeShape)
+            else:
+                if startingReservationTime + jobTimeSoFar <= targetTime:
```

Comment:
I'm a little worried that this depends on the wall times for queued jobs being correct, when they are really just guesses. If we're assuming the wall times are correct, this solution shouldn't be necessary in the first place because only the correct number of jobs will be stacked up serially in the long reservation, and the others will result in new nodes being provisioned.

Reply:
I totally agree that the correct number of jobs will be put into the long reservation, but I think the whole point of a scaler is that it tries to schedule your jobs "soon" (for some, totally arbitrary, definition of "soon"). This tries to define "soon" as 1 hour, which seems OK to me :). But, then again, it won't necessarily speed up the workflow. The time that the workflow finishes is, of course, the time that the last job finishes. So it's not necessarily best to schedule jobs earlier if you can fit them in later.

But I think it's best to schedule jobs as early as possible, because they may spawn *more* jobs--which again may spawn more jobs, etc. But you bring up a good point in that we may be wasting money by scheduling jobs than is strictly necessary.

I don't think there's a great, 100% solution to this.